lib/pbench/server/filetree.py

```diff
@@ -0,0 +1,865 @@
+
+        return Tarball(destination, controller)
+
+    def unpack(self, incoming: Path, results: Path):
+        """
+        Unpack a tarball into the INCOMING directory tree; this assumes that
+        the INCOMING controller directory already exists, which should be
+        ensured by calling this indirectly through the Controller class unpack
+        method.
+
+        TODO: This is a prototype for testing, and doesn't actually unpack
+        the tarball as we're going to be relying on the current unpack pipeline
+        command for some time.
+
+        Args:
+            incoming: Controller's directory in the INCOMING tree
+            results: Controller's directory in the RESULTS tree
+        """
+        self.logger.warning("Tarball unpack is not yet implemented")
+        unpacked = incoming / self.name
+        unpacked.mkdir(parents=True)  # Just create an empty directory for now
+        self.unpacked = unpacked
+        results_link = results / self.name
+        results_link.symlink_to(unpacked)
+        self.results_link = results_link
+
+    def uncache(self):
+        """
+        Remove the unpacked tarball directory and all contents. The Tarball
+        object isn't directly aware of the RESULTS tree at all, so the caller
+        is responsible for managing that along with the controller directories
+        in each tree.
+        """
+        if self.unpacked:
+            try:
+                shutil.rmtree(self.unpacked)
+                self.unpacked = None
+            except Exception as e:
+                self.logger.error("incoming remove for {} failed with {}", self.name, e)
+                raise
+        if self.results_link:
+            try:
+                self.results_link.unlink()
+                self.results_link = None
+            except Exception as e:
+                self.logger.error("results unlink for {} failed with {}", self.name, e)
+                raise
+
+    def delete(self):
+        """
+        Delete the tarball and MD5 file from the ARCHIVE tree.
+
+        Exceptions raised by the deletion will be propagated; nothing is
+        returned on success, but the "path" instance variables will be cleared
+        to avoid future confusion.
+        """
+        self.uncache()
+        if self.md5_path:
+            # NOTE: it's actually an error if there's no MD5 companion, but
+            # since Tarball object creation is driven by presence of the
+            # .tar.xz file we won't assume it's there to delete.
+            try:
+                self.md5_path.unlink()
+                self.md5_path = None
+            except Exception as e:
+                self.logger.error("archive unlink for {} failed with {}", self.name, e)
+                raise
+        try:
+            self.tarball_path.unlink()
+            self.tarball_path = None
+        except Exception as e:
+            self.logger.error("archive MD5 unlink for {} failed with {}", self.name, e)
+            raise
```

Comment:
If we fail to delete the MD5 file, we don't try to delete the tarball, and we don't change the `self.md5_path`.  It seems like maybe we _should_ set the path to `None`, since we're unlikely to have success next time.  And, while it seems like a bit of a long shot, why _not_ try to delete the tarball if the deletion of the hash file fails -- we came here to do that, and, since the two files are created in different ways, it's just possible that we might still be able to delete the other one.

Also, my [earlier comment](https://github.com/distributed-system-analysis/pbench/pull/2540#discussion_r763275098) on this blob of code is still unresolved.